### PR TITLE
Fortran implementation and tests of netCDF-4 performance functions

### PR DESCRIPTION
**Pull request by edhartnett**

### CODE REVIEW

File: src/clib/pio_nc.c [hunk 1]
```diff
@@ -3483,6 +3483,7 @@ int PIOc_inq_dimlen (int ncid, int dimid, PIO_Offset *lenp)
     sprintf(errstr,"in file %s",__FILE__);
   }
   ierr = check_netcdf(file, ierr, errstr,__LINE__);
+  printf("sizeof(MPI_OFFSET)=%d sizeof(PIO_Offset)=%d sizeof(*lenp)=%d \n", sizeof(MPI_OFFSET), sizeof(PIO_Offset), sizeof(*lenp));
```
Comment (jedwards4b) [inline src/clib/pio_nc.c]: is this a debug statement you meant to remove?

File: src/clib/pio_nc4.c [hunk 2]
```diff
@@ -1023,49 +844,42 @@ int PIOc_get_chunk_cache(int iotype, int io_rank, size_t *sizep,
...
-    /* if (sizep) */
-    /* 	if ((ierr = MPI_Bcast(sizep, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
-    /* if (nelemsp) */
-    /* 	if ((ierr = MPI_Bcast(nelemsp, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
-    /* if (preemptionp) */
-    /* 	if ((ierr = MPI_Bcast(preemptionp, 1, MPI_FLOAT, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
+    if (ierr)
```
Comment (jedwards4b) [inline src/clib/pio_nc4.c]: I think that this is incorrect and may cause a hang on error If some tasks set ierr but others do not.  The check_netcdf function handles this correctly.

File: src/clib/pio_nc4.c [hunk 3]
```diff
@@ -1112,6 +926,11 @@ int PIOc_set_var_chunk_cache(int ncid, int varid, size_t size, size_t nelems,
     ios = file->iosystem;
     msg = PIO_MSG_SET_VAR_CHUNK_CACHE;
 
+    int my_rank;
+    int ret;
+    if ((ret = MPI_Comm_rank(MPI_COMM_WORLD, &my_rank)))
```
Comment (jedwards4b) [inline src/clib/pio_nc4.c]: Again if MPI_Comm_rank fails on some but not all tasks - what will happen here?

File: src/clib/pio_nc4.c [hunk 4]
```diff
@@ -1206,6 +1023,11 @@ int PIOc_get_var_chunk_cache(int ncid, int varid, size_t *sizep, size_t *nelemsp
 	return PIO_EBADID;
     ios = file->iosystem;
 
+    int my_rank;
+    int ret;
+    if ((ret = MPI_Comm_rank(MPI_COMM_WORLD, &my_rank)))
```
Comment (jedwards4b) [inline src/clib/pio_nc4.c]: same as above

File: tests/unit/ncdf_tests.F90 [hunk 5]
```diff
@@ -277,29 +277,96 @@ Subroutine test_nc4(test_id, err_msg)
...
+    real :: chunk_cache_preemption_in
+
     err_msg = "no_error"
 
     dims(1) = 2*ntasks
     compdof = 2*my_rank+(/1,2/)  ! Where in the global array each task writes
     data_to_write = 1+my_rank
 
-    print*, ''
+    print*, 'PIO_initdecom'
```
Comment (jedwards4b) [inline tests/unit/ncdf_tests.F90]: spelling

**Comment by jedwards4b** [comment]
review complete - a few minor issues

File: src/clib/pio_nc.c (hunk 1 above)
Comment (edhartnett) [inline src/clib/pio_nc.c]: Good catch. Fixed.

File: src/clib/pio_nc4.c (hunk 3 above)
Comment (edhartnett) [inline src/clib/pio_nc4.c]: Those were for debugging purposes and should have been removed. They are removed now.

File: src/clib/pio_nc4.c (hunk 4 above)
Comment (edhartnett) [inline src/clib/pio_nc4.c]: Removed.

File: tests/unit/ncdf_tests.F90 (hunk 5 above)
Comment (edhartnett) [inline tests/unit/ncdf_tests.F90]: Fixed.

File: src/clib/pio_nc4.c (hunk 2 above)
Comment (edhartnett) [inline src/clib/pio_nc4.c]: I can't find this code...

PIOC_get/set_chunk_cache cannot use check_netcdf, which depends on a file being specified. The chunk cache functions are called before a file is opened, to set the chunk cache for subsequently opened files. So in those functions, I use MPI_BCAST to send the error to all processes. This is tested in the test code, and all processes do receive the error code...

Comment (jedwards4b) [inline src/clib/pio_nc4.c]: Suppose you are using NETCDF4C and the call from io-task 0 returns an error, all other tasks have ierr=PIO_NOERR,
what happens here?

Comment (edhartnett) [inline src/clib/pio_nc4.c]: OK, good question!

I have changed code to handle this better, and also added a test which
passes in a deliberately bad value for preemption to ensure that all tasks
get the resulting error code.

Please take a look at the changes and let me know what you thiink.

Thanks,
Ed

**Comment by edhartnett** [comment]
Thanks!

**Pull request merged**